// ng-zorro-antd/esm5/table/src/addon/row-indent.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: src/addon/row-indent.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var NzRowIndentDirective = /** @class */ (function () {
    function NzRowIndentDirective() {
        this.indentSize = 0;
    }
    NzRowIndentDirective.propDecorators = {
        indentSize: [{ type: Input }]
    };
NzRowIndentDirective.ɵfac = function NzRowIndentDirective_Factory(t) { return new (t || NzRowIndentDirective)(); };
NzRowIndentDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzRowIndentDirective, selectors: [["nz-row-indent"]], hostVars: 4, hostBindings: function NzRowIndentDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("padding-left", ctx.indentSize, "px");
        ɵngcc0.ɵɵclassProp("ant-table-row-indent", true);
    } }, inputs: { indentSize: "indentSize" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzRowIndentDirective, [{
        type: Directive,
        args: [{
                selector: 'nz-row-indent',
                host: {
                    '[class.ant-table-row-indent]': 'true',
                    '[style.padding-left.px]': 'indentSize'
                }
            }]
    }], function () { return []; }, { indentSize: [{
            type: Input
        }] }); })();
    return NzRowIndentDirective;
}());
export { NzRowIndentDirective };
if (false) {
    /** @type {?} */
    NzRowIndentDirective.prototype.indentSize;
}
