// ngx-nestable/esm5/lib/nestable-drag-handle/nestable-drag-handle.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { DRAG_START, REGISTER_HANDLE } from '../nestable.constant';
import * as ɵngcc0 from '@angular/core';
var NestableDragHandleDirective = /** @class */ (function () {
    function NestableDragHandleDirective(_el) {
        this._el = _el;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NestableDragHandleDirective.prototype.onMouseDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var detail = {
            param: this.ngxNestableDragHandle,
            event: event
        };
        this._el.nativeElement.dispatchEvent(new CustomEvent(DRAG_START, { bubbles: true, detail: detail }));
    };
    /**
     * @return {?}
     */
    NestableDragHandleDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._el.nativeElement.dispatchEvent(new CustomEvent(REGISTER_HANDLE, { bubbles: true, detail: this.ngxNestableDragHandle }));
    };
    /** @nocollapse */
    NestableDragHandleDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    NestableDragHandleDirective.propDecorators = {
        ngxNestableDragHandle: [{ type: Input }],
        onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
    };
NestableDragHandleDirective.ɵfac = function NestableDragHandleDirective_Factory(t) { return new (t || NestableDragHandleDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NestableDragHandleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NestableDragHandleDirective, selectors: [["", "ngxNestableDragHandle", ""]], hostBindings: function NestableDragHandleDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mousedown", function NestableDragHandleDirective_mousedown_HostBindingHandler($event) { return ctx.onMouseDown($event); });
    } }, inputs: { ngxNestableDragHandle: "ngxNestableDragHandle" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NestableDragHandleDirective, [{
        type: Directive,
        args: [{
                selector: '[ngxNestableDragHandle]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { onMouseDown: [{
            type: HostListener,
            args: ['mousedown', ['$event']]
        }], ngxNestableDragHandle: [{
            type: Input
        }] }); })();
    return NestableDragHandleDirective;
}());
export { NestableDragHandleDirective };
if (false) {
    /** @type {?} */
    NestableDragHandleDirective.prototype.ngxNestableDragHandle;
    /**
     * @type {?}
     * @private
     */
    NestableDragHandleDirective.prototype._el;
}
