// ng-chartist/esm5/lib/chartist.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ChartistComponent } from './chartist.component';
import * as ɵngcc0 from '@angular/core';
var ChartistModule = /** @class */ (function () {
    function ChartistModule() {
    }
ChartistModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ChartistModule });
ChartistModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ChartistModule_Factory(t) { return new (t || ChartistModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ChartistModule, { declarations: function () { return [ChartistComponent]; }, exports: function () { return [ChartistComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ChartistModule, [{
        type: NgModule,
        args: [{
                declarations: [ChartistComponent],
                imports: [],
                exports: [ChartistComponent]
            }]
    }], function () { return []; }, null); })();
    return ChartistModule;
}());
export { ChartistModule };
