// ng-zorro-antd/esm5/slider/track.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: track.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __decorate, __metadata } from "tslib";
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { InputBoolean, InputNumber } from 'ng-zorro-antd/core/util';
/**
 * @record
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
export function NzSliderTrackStyle() { }
if (false) {
    /** @type {?|undefined} */
    NzSliderTrackStyle.prototype.bottom;
    /** @type {?|undefined} */
    NzSliderTrackStyle.prototype.height;
    /** @type {?|undefined} */
    NzSliderTrackStyle.prototype.left;
    /** @type {?|undefined} */
    NzSliderTrackStyle.prototype.width;
    /** @type {?|undefined} */
    NzSliderTrackStyle.prototype.visibility;
}
var NzSliderTrackComponent = /** @class */ (function () {
    function NzSliderTrackComponent() {
        this.offset = 0;
        this.reverse = false;
        this.length = 0;
        this.vertical = false;
        this.included = false;
        this.style = {};
    }
    /**
     * @return {?}
     */
    NzSliderTrackComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        var _a, _b;
        /** @type {?} */
        var vertical = this.vertical;
        /** @type {?} */
        var reverse = this.reverse;
        /** @type {?} */
        var visibility = this.included ? 'visible' : 'hidden';
        /** @type {?} */
        var offset = this.offset;
        /** @type {?} */
        var length = this.length;
        /** @type {?} */
        var positonStyle = vertical
            ? (_a = {},
                _a[reverse ? 'top' : 'bottom'] = offset + "%",
                _a[reverse ? 'bottom' : 'top'] = 'auto',
                _a.height = length + "%",
                _a.visibility = visibility,
                _a) : (_b = {},
            _b[reverse ? 'right' : 'left'] = offset + "%",
            _b[reverse ? 'left' : 'right'] = 'auto',
            _b.width = length + "%",
            _b.visibility = visibility,
            _b);
        this.style = positonStyle;
    };
    NzSliderTrackComponent.propDecorators = {
        offset: [{ type: Input }],
        reverse: [{ type: Input }],
        length: [{ type: Input }],
        vertical: [{ type: Input }],
        included: [{ type: Input }]
    };
    __decorate([
        InputNumber(),
        __metadata("design:type", Number)
    ], NzSliderTrackComponent.prototype, "offset", void 0);
    __decorate([
        InputBoolean(),
        __metadata("design:type", Boolean)
    ], NzSliderTrackComponent.prototype, "reverse", void 0);
    __decorate([
        InputNumber(),
        __metadata("design:type", Number)
    ], NzSliderTrackComponent.prototype, "length", void 0);
    __decorate([
        InputBoolean(),
        __metadata("design:type", Object)
    ], NzSliderTrackComponent.prototype, "vertical", void 0);
    __decorate([
        InputBoolean(),
        __metadata("design:type", Object)
    ], NzSliderTrackComponent.prototype, "included", void 0);
NzSliderTrackComponent.ɵfac = function NzSliderTrackComponent_Factory(t) { return new (t || NzSliderTrackComponent)(); };
NzSliderTrackComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NzSliderTrackComponent, selectors: [["nz-slider-track"]], inputs: { offset: "offset", reverse: "reverse", length: "length", vertical: "vertical", included: "included" }, exportAs: ["nzSliderTrack"], features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 1, vars: 1, consts: [[1, "ant-slider-track", 3, "ngStyle"]], template: function NzSliderTrackComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngStyle", ctx.style);
    } }, directives: [ɵngcc1.NgStyle], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzSliderTrackComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'nz-slider-track',
                exportAs: 'nzSliderTrack',
                preserveWhitespaces: false,
                template: " <div class=\"ant-slider-track\" [ngStyle]=\"style\"></div> "
            }]
    }], function () { return []; }, { offset: [{
            type: Input
        }], reverse: [{
            type: Input
        }], length: [{
            type: Input
        }], vertical: [{
            type: Input
        }], included: [{
            type: Input
        }] }); })();
    return NzSliderTrackComponent;
}());
export { NzSliderTrackComponent };
if (false) {
    /** @type {?} */
    NzSliderTrackComponent.ngAcceptInputType_offset;
    /** @type {?} */
    NzSliderTrackComponent.ngAcceptInputType_length;
    /** @type {?} */
    NzSliderTrackComponent.ngAcceptInputType_vertical;
    /** @type {?} */
    NzSliderTrackComponent.ngAcceptInputType_included;
    /** @type {?} */
    NzSliderTrackComponent.ngAcceptInputType_reverse;
    /** @type {?} */
    NzSliderTrackComponent.prototype.offset;
    /** @type {?} */
    NzSliderTrackComponent.prototype.reverse;
    /** @type {?} */
    NzSliderTrackComponent.prototype.length;
    /** @type {?} */
    NzSliderTrackComponent.prototype.vertical;
    /** @type {?} */
    NzSliderTrackComponent.prototype.included;
    /** @type {?} */
    NzSliderTrackComponent.prototype.style;
}
