// @ngx-progressbar/router/esm5/lib/ng-progress-router.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NG_PROGRESS_ROUTER_CONFIG } from './ng-progress-router.interface';
import { NgProgressRouter } from './ng-progress-router.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './ng-progress-router.service';
var NgProgressRouterModule = /** @class */ (function () {
    // Inject the service to activate it
    function NgProgressRouterModule(ngProgressRouter) {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    NgProgressRouterModule.withConfig = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: NgProgressRouterModule,
            providers: [
                { provide: NG_PROGRESS_ROUTER_CONFIG, useValue: config }
            ]
        };
    };
    /** @nocollapse */
    NgProgressRouterModule.ctorParameters = function () { return [
        { type: NgProgressRouter }
    ]; };
NgProgressRouterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgProgressRouterModule });
NgProgressRouterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgProgressRouterModule_Factory(t) { return new (t || NgProgressRouterModule)(ɵngcc0.ɵɵinject(ɵngcc1.NgProgressRouter)); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgProgressRouterModule, [{
        type: NgModule,
        args: [{}]
    }], function () { return [{ type: ɵngcc1.NgProgressRouter }]; }, null); })();
    return NgProgressRouterModule;
}());
export { NgProgressRouterModule };
