// ngx-nestable/esm2015/lib/nestable-expand-collapse-handle/nestable-expand-collapse.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { EXPAND_COLLAPSE } from '../nestable.constant';
import * as ɵngcc0 from '@angular/core';
export class NestableExpandCollapseDirective {
    /**
     * @param {?} _el
     */
    constructor(_el) {
        this._el = _el;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseDown(event) {
        event.stopPropagation();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        this.ngxNestableExpandCollapse.item['$$expanded'] = !this.ngxNestableExpandCollapse.item['$$expanded'];
        this._el.nativeElement.dispatchEvent(new CustomEvent(EXPAND_COLLAPSE, {
            bubbles: true,
            detail: this.ngxNestableExpandCollapse
        }));
    }
}
NestableExpandCollapseDirective.ɵfac = function NestableExpandCollapseDirective_Factory(t) { return new (t || NestableExpandCollapseDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NestableExpandCollapseDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NestableExpandCollapseDirective, selectors: [["", "ngxNestableExpandCollapse", ""]], hostBindings: function NestableExpandCollapseDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mousedown", function NestableExpandCollapseDirective_mousedown_HostBindingHandler($event) { return ctx.onMouseDown($event); })("click", function NestableExpandCollapseDirective_click_HostBindingHandler($event) { return ctx.onClick($event); });
    } }, inputs: { ngxNestableExpandCollapse: "ngxNestableExpandCollapse" } });
/** @nocollapse */
NestableExpandCollapseDirective.ctorParameters = () => [
    { type: ElementRef }
];
NestableExpandCollapseDirective.propDecorators = {
    ngxNestableExpandCollapse: [{ type: Input }],
    onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NestableExpandCollapseDirective, [{
        type: Directive,
        args: [{
                selector: '[ngxNestableExpandCollapse]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { 
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseDown: [{
            type: HostListener,
            args: ['mousedown', ['$event']]
        }], 
    /**
     * @param {?} event
     * @return {?}
     */
    onClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }], ngxNestableExpandCollapse: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    NestableExpandCollapseDirective.prototype.ngxNestableExpandCollapse;
    /**
     * @type {?}
     * @private
     */
    NestableExpandCollapseDirective.prototype._el;
}
