// ngx-nestable/esm5/lib/nestable-expand-collapse-handle/nestable-expand-collapse.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { EXPAND_COLLAPSE } from '../nestable.constant';
import * as ɵngcc0 from '@angular/core';
var NestableExpandCollapseDirective = /** @class */ (function () {
    function NestableExpandCollapseDirective(_el) {
        this._el = _el;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NestableExpandCollapseDirective.prototype.onMouseDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NestableExpandCollapseDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.ngxNestableExpandCollapse.item['$$expanded'] = !this.ngxNestableExpandCollapse.item['$$expanded'];
        this._el.nativeElement.dispatchEvent(new CustomEvent(EXPAND_COLLAPSE, {
            bubbles: true,
            detail: this.ngxNestableExpandCollapse
        }));
    };
    /** @nocollapse */
    NestableExpandCollapseDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    NestableExpandCollapseDirective.propDecorators = {
        ngxNestableExpandCollapse: [{ type: Input }],
        onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
NestableExpandCollapseDirective.ɵfac = function NestableExpandCollapseDirective_Factory(t) { return new (t || NestableExpandCollapseDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NestableExpandCollapseDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NestableExpandCollapseDirective, selectors: [["", "ngxNestableExpandCollapse", ""]], hostBindings: function NestableExpandCollapseDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mousedown", function NestableExpandCollapseDirective_mousedown_HostBindingHandler($event) { return ctx.onMouseDown($event); })("click", function NestableExpandCollapseDirective_click_HostBindingHandler($event) { return ctx.onClick($event); });
    } }, inputs: { ngxNestableExpandCollapse: "ngxNestableExpandCollapse" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NestableExpandCollapseDirective, [{
        type: Directive,
        args: [{
                selector: '[ngxNestableExpandCollapse]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { onMouseDown: [{
            type: HostListener,
            args: ['mousedown', ['$event']]
        }], onClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }], ngxNestableExpandCollapse: [{
            type: Input
        }] }); })();
    return NestableExpandCollapseDirective;
}());
export { NestableExpandCollapseDirective };
if (false) {
    /** @type {?} */
    NestableExpandCollapseDirective.prototype.ngxNestableExpandCollapse;
    /**
     * @type {?}
     * @private
     */
    NestableExpandCollapseDirective.prototype._el;
}
