// angular-us-map/us-map.component.js
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { UsMapService } from './us-map.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './us-map.service';
import * as ɵngcc2 from '@angular/common';

function UsMapComponent__svg_path_11_Template(rf, ctx) { if (rf & 1) {
    var _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "path", 73);
    ɵngcc0.ɵɵlistener("click", function UsMapComponent__svg_path_11_Template__svg_path_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r4); var coordinate_r1 = ctx.$implicit; var ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onUsMapClick(coordinate_r1.id); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var coordinate_r1 = ctx.$implicit;
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("id", coordinate_r1.id)("d", coordinate_r1.d)("fill", ctx_r0.fillStateColor)("stroke", ctx_r0.strokeColor);
} }
var UsMapComponent = (function () {
    function UsMapComponent(usMapService) {
        this.usMapService = usMapService;
        this.fillColor = "#FFFFFF";
        this.fillStateColor = "#FF9900";
        this.strokeColor = "#000000";
        this.click = new EventEmitter();
    }
    UsMapComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.usMapService.getUsMapCoordinates().then(function (data) { return _this.coordinates = data; });
    };
    UsMapComponent.prototype.onUsMapClick = function (state) {
        this.click.emit({ "state-abbr": state });
    };
    /** @nocollapse */
    UsMapComponent.ctorParameters = function () { return [
        { type: UsMapService, },
    ]; };
    UsMapComponent.propDecorators = {
        'fillColor': [{ type: Input },],
        'fillStateColor': [{ type: Input },],
        'strokeColor': [{ type: Input },],
        'click': [{ type: Output, args: ['onMapClick',] },],
    };
UsMapComponent.ɵfac = function UsMapComponent_Factory(t) { return new (t || UsMapComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UsMapService)); };
UsMapComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: UsMapComponent, selectors: [["us-map"]], inputs: { fillColor: "fillColor", fillStateColor: "fillStateColor", strokeColor: "strokeColor" }, outputs: { click: "onMapClick" }, decls: 173, vars: 69, consts: [["xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 1000 600"], ["x", "895", "y", "130", "width", "30", "height", "17", "id", "MA_1", 3, "click"], ["x", "890", "y", "182", "width", "21", "height", "17", "id", "RI_1", 3, "click"], ["x", "869", "y", "196", "width", "21", "height", "18", "id", "CT_1", 3, "click"], ["x", "851", "y", "216", "width", "21", "height", "18", "id", "NJ_1", 3, "click"], ["x", "843", "y", "240", "width", "24", "height", "18", "id", "DE_1", 3, "click"], ["x", "843", "y", "261", "width", "30", "height", "18", "id", "MD_1", 3, "click"], ["x", "817", "y", "82", "width", "27", "height", "18", "id", "VT_1", 3, "click"], ["x", "837", "y", "66", "width", "27", "height", "18", "id", "NH_1", 3, "click"], ["x", "361", "y", "562", "width", "24", "height", "18", "id", "HI_1", 3, "click"], [3, "click", 4, "ngFor", "ngForOf"], ["x", "657", "y", "405", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["dy", "6.5"], ["x", "115", "y", "490", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "205", "y", "355", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "547", "y", "367", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "70", "y", "260", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "322", "y", "273", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "880", "y", "205", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "855", "y", "248", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "765", "y", "500", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "715", "y", "400", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "374", "y", "570", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "195", "y", "145", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "598", "y", "253", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "645", "y", "253", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "525", "y", "213", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "445", "y", "290", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "675", "y", "300", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "549", "y", "447", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "892", "y", "80", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "858", "y", "270", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "910", "y", "137", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "658", "y", "174", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "505", "y", "127", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "603", "y", "410", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "540", "y", "288", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "287", "y", "90", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "424", "y", "226", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "140", "y", "235", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "850", "y", "75", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "860", "y", "224", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "304", "y", "369", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "816", "y", "151", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "787", "y", "327", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "420", "y", "97", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "700", "y", "240", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "460", "y", "358", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "103", "y", "123", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "783", "y", "210", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "900", "y", "190", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "761", "y", "370", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "420", "y", "161", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "656", "y", "339", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "429", "y", "450", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "224", "y", "254", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "830", "y", "90", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "788", "y", "282", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "125", "y", "50", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "742", "y", "269", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "578", "y", "156", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["x", "300", "y", "183", "text-anchor", "middle", "stroke", "none", 2, "cursor", "pointer", 3, "click"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "830", "y1", "93", "x2", "845", "y2", "120"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "850", "y1", "78", "x2", "865", "y2", "120"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "898", "y1", "140", "x2", "875", "y2", "155"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "892", "y1", "188", "x2", "878", "y2", "173"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "880", "y1", "198", "x2", "858", "y2", "175"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "840", "y1", "215", "x2", "850", "y2", "220"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "828", "y1", "248", "x2", "842", "y2", "249"], ["xmlns", "http://www.w3.org/2000/svg", "fill", "none", "stroke", "#C0C0C0", "stroke-width", "2", "x1", "802", "y1", "245", "x2", "845", "y2", "269"], ["id", "DC"], ["id", "DC1", "fill", "#D3D3D3", "d", "M801.8,253.8 l-1.1-1.6 -1-0.8 1.1-1.6 2.2,1.5z"], ["id", "DC2", "fill", "#D3D3D3", "stroke", "#FFFFFF", "stroke-width", "1.5", "cx", "801.3", "cy", "251.8", "r", "5", "opacity", "0"], [3, "click"]], template: function UsMapComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵnamespaceSVG();
        ɵngcc0.ɵɵelementStart(0, "svg", 0);
        ɵngcc0.ɵɵelementStart(1, "rect", 1);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_1_listener() { return ctx.onUsMapClick("MA"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(2, "rect", 2);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_2_listener() { return ctx.onUsMapClick("RI"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "rect", 3);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_3_listener() { return ctx.onUsMapClick("CT"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "rect", 4);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_4_listener() { return ctx.onUsMapClick("NJ"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "rect", 5);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_5_listener() { return ctx.onUsMapClick("DE"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "rect", 6);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_6_listener() { return ctx.onUsMapClick("MD"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "rect", 7);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_7_listener() { return ctx.onUsMapClick("VT"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "rect", 8);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_8_listener() { return ctx.onUsMapClick("NH"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "rect", 9);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_rect_click_9_listener() { return ctx.onUsMapClick("HI"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "g");
        ɵngcc0.ɵɵtemplate(11, UsMapComponent__svg_path_11_Template, 1, 4, "path", 10);
        ɵngcc0.ɵɵelementStart(12, "text", 11);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_12_listener() { return ctx.onUsMapClick("AL"); });
        ɵngcc0.ɵɵelementStart(13, "tspan", 12);
        ɵngcc0.ɵɵtext(14, "AL");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "text", 13);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_15_listener() { return ctx.onUsMapClick("AK"); });
        ɵngcc0.ɵɵelementStart(16, "tspan", 12);
        ɵngcc0.ɵɵtext(17, "AK");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "text", 14);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_18_listener() { return ctx.onUsMapClick("AZ"); });
        ɵngcc0.ɵɵelementStart(19, "tspan", 12);
        ɵngcc0.ɵɵtext(20, "AZ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "text", 15);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_21_listener() { return ctx.onUsMapClick("AR"); });
        ɵngcc0.ɵɵelementStart(22, "tspan", 12);
        ɵngcc0.ɵɵtext(23, "AR");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "text", 16);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_24_listener() { return ctx.onUsMapClick("CA"); });
        ɵngcc0.ɵɵelementStart(25, "tspan", 12);
        ɵngcc0.ɵɵtext(26, "CA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "text", 17);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_27_listener() { return ctx.onUsMapClick("CO"); });
        ɵngcc0.ɵɵelementStart(28, "tspan", 12);
        ɵngcc0.ɵɵtext(29, "CO");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(30, "text", 18);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_30_listener() { return ctx.onUsMapClick("CT"); });
        ɵngcc0.ɵɵelementStart(31, "tspan", 12);
        ɵngcc0.ɵɵtext(32, "CT");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(33, "text", 19);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_33_listener() { return ctx.onUsMapClick("DE"); });
        ɵngcc0.ɵɵelementStart(34, "tspan", 12);
        ɵngcc0.ɵɵtext(35, "DE");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(36, "text", 20);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_36_listener() { return ctx.onUsMapClick("FL"); });
        ɵngcc0.ɵɵelementStart(37, "tspan", 12);
        ɵngcc0.ɵɵtext(38, "FL");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "text", 21);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_39_listener() { return ctx.onUsMapClick("GA"); });
        ɵngcc0.ɵɵelementStart(40, "tspan", 12);
        ɵngcc0.ɵɵtext(41, "GA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "text", 22);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_42_listener() { return ctx.onUsMapClick("HI"); });
        ɵngcc0.ɵɵelementStart(43, "tspan", 12);
        ɵngcc0.ɵɵtext(44, "HI");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(45, "text", 23);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_45_listener() { return ctx.onUsMapClick("ID"); });
        ɵngcc0.ɵɵelementStart(46, "tspan", 12);
        ɵngcc0.ɵɵtext(47, "ID");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(48, "text", 24);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_48_listener() { return ctx.onUsMapClick("IL"); });
        ɵngcc0.ɵɵelementStart(49, "tspan", 12);
        ɵngcc0.ɵɵtext(50, "IL");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(51, "text", 25);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_51_listener() { return ctx.onUsMapClick("IN"); });
        ɵngcc0.ɵɵelementStart(52, "tspan", 12);
        ɵngcc0.ɵɵtext(53, "IN");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(54, "text", 26);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_54_listener() { return ctx.onUsMapClick("IA"); });
        ɵngcc0.ɵɵelementStart(55, "tspan", 12);
        ɵngcc0.ɵɵtext(56, "IA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(57, "text", 27);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_57_listener() { return ctx.onUsMapClick("KS"); });
        ɵngcc0.ɵɵelementStart(58, "tspan", 12);
        ɵngcc0.ɵɵtext(59, "KS");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(60, "text", 28);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_60_listener() { return ctx.onUsMapClick("KY"); });
        ɵngcc0.ɵɵelementStart(61, "tspan", 12);
        ɵngcc0.ɵɵtext(62, "KY");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(63, "text", 29);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_63_listener() { return ctx.onUsMapClick("LA"); });
        ɵngcc0.ɵɵelementStart(64, "tspan", 12);
        ɵngcc0.ɵɵtext(65, "LA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(66, "text", 30);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_66_listener() { return ctx.onUsMapClick("ME"); });
        ɵngcc0.ɵɵelementStart(67, "tspan", 12);
        ɵngcc0.ɵɵtext(68, "ME");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(69, "text", 31);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_69_listener() { return ctx.onUsMapClick("MD"); });
        ɵngcc0.ɵɵelementStart(70, "tspan", 12);
        ɵngcc0.ɵɵtext(71, "MD");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(72, "text", 32);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_72_listener() { return ctx.onUsMapClick("MA"); });
        ɵngcc0.ɵɵelementStart(73, "tspan", 12);
        ɵngcc0.ɵɵtext(74, "MA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(75, "text", 33);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_75_listener() { return ctx.onUsMapClick("MI"); });
        ɵngcc0.ɵɵelementStart(76, "tspan", 12);
        ɵngcc0.ɵɵtext(77, "MI");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(78, "text", 34);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_78_listener() { return ctx.onUsMapClick("MN"); });
        ɵngcc0.ɵɵelementStart(79, "tspan", 12);
        ɵngcc0.ɵɵtext(80, "MN");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(81, "text", 35);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_81_listener() { return ctx.onUsMapClick("MS"); });
        ɵngcc0.ɵɵelementStart(82, "tspan", 12);
        ɵngcc0.ɵɵtext(83, "MS");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(84, "text", 36);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_84_listener() { return ctx.onUsMapClick("MO"); });
        ɵngcc0.ɵɵelementStart(85, "tspan", 12);
        ɵngcc0.ɵɵtext(86, "MO");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(87, "text", 37);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_87_listener() { return ctx.onUsMapClick("MT"); });
        ɵngcc0.ɵɵelementStart(88, "tspan", 12);
        ɵngcc0.ɵɵtext(89, "MT");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(90, "text", 38);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_90_listener() { return ctx.onUsMapClick("NE"); });
        ɵngcc0.ɵɵelementStart(91, "tspan", 12);
        ɵngcc0.ɵɵtext(92, "NE");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(93, "text", 39);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_93_listener() { return ctx.onUsMapClick("NV"); });
        ɵngcc0.ɵɵelementStart(94, "tspan", 12);
        ɵngcc0.ɵɵtext(95, "NV");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(96, "text", 40);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_96_listener() { return ctx.onUsMapClick("NH"); });
        ɵngcc0.ɵɵelementStart(97, "tspan", 12);
        ɵngcc0.ɵɵtext(98, "NH");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(99, "text", 41);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_99_listener() { return ctx.onUsMapClick("NJ"); });
        ɵngcc0.ɵɵelementStart(100, "tspan", 12);
        ɵngcc0.ɵɵtext(101, "NJ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(102, "text", 42);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_102_listener() { return ctx.onUsMapClick("NM"); });
        ɵngcc0.ɵɵelementStart(103, "tspan", 12);
        ɵngcc0.ɵɵtext(104, "NM");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(105, "text", 43);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_105_listener() { return ctx.onUsMapClick("NY"); });
        ɵngcc0.ɵɵelementStart(106, "tspan", 12);
        ɵngcc0.ɵɵtext(107, "NY");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(108, "text", 44);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_108_listener() { return ctx.onUsMapClick("NC"); });
        ɵngcc0.ɵɵelementStart(109, "tspan", 12);
        ɵngcc0.ɵɵtext(110, "NC");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(111, "text", 45);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_111_listener() { return ctx.onUsMapClick("ND"); });
        ɵngcc0.ɵɵelementStart(112, "tspan", 12);
        ɵngcc0.ɵɵtext(113, "ND");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(114, "text", 46);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_114_listener() { return ctx.onUsMapClick("OH"); });
        ɵngcc0.ɵɵelementStart(115, "tspan", 12);
        ɵngcc0.ɵɵtext(116, "OH");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(117, "text", 47);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_117_listener() { return ctx.onUsMapClick("OK"); });
        ɵngcc0.ɵɵelementStart(118, "tspan", 12);
        ɵngcc0.ɵɵtext(119, "OK");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(120, "text", 48);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_120_listener() { return ctx.onUsMapClick("OR"); });
        ɵngcc0.ɵɵelementStart(121, "tspan", 12);
        ɵngcc0.ɵɵtext(122, "OR");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(123, "text", 49);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_123_listener() { return ctx.onUsMapClick("PA"); });
        ɵngcc0.ɵɵelementStart(124, "tspan", 12);
        ɵngcc0.ɵɵtext(125, "PA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(126, "text", 50);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_126_listener() { return ctx.onUsMapClick("RI"); });
        ɵngcc0.ɵɵelementStart(127, "tspan", 12);
        ɵngcc0.ɵɵtext(128, "RI");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(129, "text", 51);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_129_listener() { return ctx.onUsMapClick("SC"); });
        ɵngcc0.ɵɵelementStart(130, "tspan", 12);
        ɵngcc0.ɵɵtext(131, "SC");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(132, "text", 52);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_132_listener() { return ctx.onUsMapClick("SD"); });
        ɵngcc0.ɵɵelementStart(133, "tspan", 12);
        ɵngcc0.ɵɵtext(134, "SD");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(135, "text", 53);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_135_listener() { return ctx.onUsMapClick("TN"); });
        ɵngcc0.ɵɵelementStart(136, "tspan", 12);
        ɵngcc0.ɵɵtext(137, "TN");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(138, "text", 54);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_138_listener() { return ctx.onUsMapClick("TX"); });
        ɵngcc0.ɵɵelementStart(139, "tspan", 12);
        ɵngcc0.ɵɵtext(140, "TX");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(141, "text", 55);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_141_listener() { return ctx.onUsMapClick("UT"); });
        ɵngcc0.ɵɵelementStart(142, "tspan", 12);
        ɵngcc0.ɵɵtext(143, "UT");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(144, "text", 56);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_144_listener() { return ctx.onUsMapClick("VT"); });
        ɵngcc0.ɵɵelementStart(145, "tspan", 12);
        ɵngcc0.ɵɵtext(146, "VT");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(147, "text", 57);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_147_listener() { return ctx.onUsMapClick("VA"); });
        ɵngcc0.ɵɵelementStart(148, "tspan", 12);
        ɵngcc0.ɵɵtext(149, "VA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(150, "text", 58);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_150_listener() { return ctx.onUsMapClick("WA"); });
        ɵngcc0.ɵɵelementStart(151, "tspan", 12);
        ɵngcc0.ɵɵtext(152, "WA");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(153, "text", 59);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_153_listener() { return ctx.onUsMapClick("WV"); });
        ɵngcc0.ɵɵelementStart(154, "tspan", 12);
        ɵngcc0.ɵɵtext(155, "WV");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(156, "text", 60);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_156_listener() { return ctx.onUsMapClick("WI"); });
        ɵngcc0.ɵɵelementStart(157, "tspan", 12);
        ɵngcc0.ɵɵtext(158, "WI");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(159, "text", 61);
        ɵngcc0.ɵɵlistener("click", function UsMapComponent_Template__svg_text_click_159_listener() { return ctx.onUsMapClick("WY"); });
        ɵngcc0.ɵɵelementStart(160, "tspan", 12);
        ɵngcc0.ɵɵtext(161, "WY");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(162, "line", 62);
        ɵngcc0.ɵɵelement(163, "line", 63);
        ɵngcc0.ɵɵelement(164, "line", 64);
        ɵngcc0.ɵɵelement(165, "line", 65);
        ɵngcc0.ɵɵelement(166, "line", 66);
        ɵngcc0.ɵɵelement(167, "line", 67);
        ɵngcc0.ɵɵelement(168, "line", 68);
        ɵngcc0.ɵɵelement(169, "line", 69);
        ɵngcc0.ɵɵelementStart(170, "g", 70);
        ɵngcc0.ɵɵelement(171, "path", 71);
        ɵngcc0.ɵɵelement(172, "circle", 72);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillStateColor)("stroke", ctx.strokeColor);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.coordinates);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("fill", ctx.fillColor);
    } }, directives: [ɵngcc2.NgForOf], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(UsMapComponent, [{
        type: Component,
        args: [{
                selector: 'us-map',
                template: "\n\t\t<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 1000 600\">\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"895\" y=\"130\" width=\"30\" height=\"17\" id=\"MA_1\" (click)=\"onUsMapClick('MA')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"890\" y=\"182\" width=\"21\" height=\"17\" id=\"RI_1\" (click)=\"onUsMapClick('RI')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"869\" y=\"196\" width=\"21\" height=\"18\" id=\"CT_1\" (click)=\"onUsMapClick('CT')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"851\" y=\"216\" width=\"21\" height=\"18\" id=\"NJ_1\" (click)=\"onUsMapClick('NJ')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"843\" y=\"240\" width=\"24\" height=\"18\" id=\"DE_1\" (click)=\"onUsMapClick('DE')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"843\" y=\"261\" width=\"30\" height=\"18\" id=\"MD_1\" (click)=\"onUsMapClick('MD')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"817\" y=\"82\" width=\"27\" height=\"18\" id=\"VT_1\" (click)=\"onUsMapClick('VT')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"837\" y=\"66\" width=\"27\" height=\"18\" id=\"NH_1\" (click)=\"onUsMapClick('NH')\"/>\n\t\t    <rect [attr.fill]=\"fillStateColor\" [attr.stroke]=\"strokeColor\" x=\"361\" y=\"562\" width=\"24\" height=\"18\" id=\"HI_1\" (click)=\"onUsMapClick('HI')\"/>\n\t\t    <g>\n\t\t        <path \n\t\t            *ngFor=\"let coordinate of coordinates;let i = index\"\n\t\t            [attr.id]=\"coordinate.id\"\n\t\t            [attr.d]=\"coordinate.d\"\n\t\t            [attr.fill]=\"fillStateColor\"\n\t\t            [attr.stroke]=\"strokeColor\"            \n\t\t            (click)=\"onUsMapClick(coordinate.id)\"\n\t\t        >\n\t\t        </path>\n\t\t        <text style=\"cursor: pointer;\" x=\"657\" y=\"405\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('AL')\"><tspan dy=\"6.5\">AL</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"115\" y=\"490\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('AK')\"><tspan dy=\"6.5\">AK</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"205\" y=\"355\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('AZ')\"><tspan dy=\"6.5\">AZ</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"547\" y=\"367\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('AR')\"><tspan dy=\"6.5\">AR</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"70\" y=\"260\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('CA')\"><tspan dy=\"6.5\">CA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"322\" y=\"273\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('CO')\"><tspan dy=\"6.5\">CO</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"880\" y=\"205\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('CT')\"><tspan dy=\"6.5\">CT</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"855\" y=\"248\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('DE')\"><tspan dy=\"6.5\">DE</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"765\" y=\"500\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('FL')\"><tspan dy=\"6.5\">FL</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"715\" y=\"400\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('GA')\"><tspan dy=\"6.5\">GA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"374\" y=\"570\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('HI')\"><tspan dy=\"6.5\">HI</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"195\" y=\"145\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('ID')\"><tspan dy=\"6.5\">ID</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"598\" y=\"253\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('IL')\"><tspan dy=\"6.5\">IL</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"645\" y=\"253\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('IN')\"><tspan dy=\"6.5\">IN</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"525\" y=\"213\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('IA')\"><tspan dy=\"6.5\">IA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"445\" y=\"290\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('KS')\"><tspan dy=\"6.5\">KS</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"675\" y=\"300\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('KY')\"><tspan dy=\"6.5\">KY</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"549\" y=\"447\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('LA')\"><tspan dy=\"6.5\">LA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"892\" y=\"80\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('ME')\"><tspan dy=\"6.5\">ME</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"858\" y=\"270\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('MD')\"><tspan dy=\"6.5\">MD</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"910\" y=\"137\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('MA')\"><tspan dy=\"6.5\">MA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"658\" y=\"174\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('MI')\"><tspan dy=\"6.5\">MI</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"505\" y=\"127\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('MN')\"><tspan dy=\"6.5\">MN</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"603\" y=\"410\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('MS')\"><tspan dy=\"6.5\">MS</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"540\" y=\"288\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('MO')\"><tspan dy=\"6.5\">MO</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"287\" y=\"90\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('MT')\"><tspan dy=\"6.5\">MT</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"424\" y=\"226\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('NE')\"><tspan dy=\"6.5\">NE</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"140\" y=\"235\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('NV')\"><tspan dy=\"6.5\">NV</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"850\" y=\"75\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('NH')\"><tspan dy=\"6.5\">NH</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"860\" y=\"224\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('NJ')\"><tspan dy=\"6.5\">NJ</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"304\" y=\"369\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('NM')\"><tspan dy=\"6.5\">NM</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"816\" y=\"151\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('NY')\"><tspan dy=\"6.5\">NY</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"787\" y=\"327\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('NC')\"><tspan dy=\"6.5\">NC</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"420\" y=\"97\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('ND')\"><tspan dy=\"6.5\">ND</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"700\" y=\"240\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('OH')\"><tspan dy=\"6.5\">OH</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"460\" y=\"358\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('OK')\"><tspan dy=\"6.5\">OK</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"103\" y=\"123\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('OR')\"><tspan dy=\"6.5\">OR</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"783\" y=\"210\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('PA')\"><tspan dy=\"6.5\">PA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"900\" y=\"190\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('RI')\"><tspan dy=\"6.5\">RI</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"761\" y=\"370\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('SC')\"><tspan dy=\"6.5\">SC</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"420\" y=\"161\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('SD')\"><tspan dy=\"6.5\">SD</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"656\" y=\"339\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('TN')\"><tspan dy=\"6.5\">TN</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"429\" y=\"450\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('TX')\"><tspan dy=\"6.5\">TX</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"224\" y=\"254\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('UT')\"><tspan dy=\"6.5\">UT</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"830\" y=\"90\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('VT')\"><tspan dy=\"6.5\">VT</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"788\" y=\"282\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('VA')\"><tspan dy=\"6.5\">VA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"125\" y=\"50\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('WA')\"><tspan dy=\"6.5\">WA</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"742\" y=\"269\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('WV')\"><tspan dy=\"6.5\">WV</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"578\" y=\"156\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('WI')\"><tspan dy=\"6.5\">WI</tspan></text>\n\t\t        <text style=\"cursor: pointer;\" x=\"300\" y=\"183\" text-anchor=\"middle\" stroke=\"none\" [attr.fill]=\"fillColor\" (click)=\"onUsMapClick('WY')\"><tspan dy=\"6.5\">WY</tspan></text>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"830\" y1=\"93\" x2=\"845\" y2=\"120\"/>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"850\" y1=\"78\" x2=\"865\" y2=\"120\"/>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"898\" y1=\"140\" x2=\"875\" y2=\"155\"/>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"892\" y1=\"188\" x2=\"878\" y2=\"173\"/>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"880\" y1=\"198\" x2=\"858\" y2=\"175\"/>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"840\" y1=\"215\" x2=\"850\" y2=\"220\"/>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"828\" y1=\"248\" x2=\"842\" y2=\"249\"/>\n\t\t        <line xmlns=\"http://www.w3.org/2000/svg\" fill=\"none\" stroke=\"#C0C0C0\" stroke-width=\"2\" x1=\"802\" y1=\"245\" x2=\"845\" y2=\"269\"/>\n\t\t        <g id=\"DC\">\n\t\t            <path id=\"DC1\" fill=\"#D3D3D3\" d=\"M801.8,253.8 l-1.1-1.6 -1-0.8 1.1-1.6 2.2,1.5z\"/>\n\t\t            <circle id=\"DC2\" fill=\"#D3D3D3\" stroke=\"#FFFFFF\" stroke-width=\"1.5\" cx=\"801.3\" cy=\"251.8\" r=\"5\" opacity=\"0\"/> <!-- Set opacity to \"0\" to hide DC circle -->\n\t\t        </g>        \n\t\t    </g>\n\t\t</svg>\n\t"
            }]
    }], function () { return [{ type: ɵngcc1.UsMapService }]; }, { fillColor: [{
            type: Input
        }], fillStateColor: [{
            type: Input
        }], strokeColor: [{
            type: Input
        }], click: [{
            type: Output,
            args: ['onMapClick']
        }] }); })();
    return UsMapComponent;
}());
export { UsMapComponent };
