// angular-us-map/us-map.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { UsMapService } from './us-map.service';
import { UsMapComponent } from './us-map.component';
import * as ɵngcc0 from '@angular/core';
var UsMapModule = (function () {
    function UsMapModule() {
    }
    /** @nocollapse */
    UsMapModule.ctorParameters = function () { return []; };
UsMapModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: UsMapModule });
UsMapModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function UsMapModule_Factory(t) { return new (t || UsMapModule)(); }, providers: [UsMapService], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(UsMapModule, { declarations: function () { return [UsMapComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [UsMapComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(UsMapModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [UsMapComponent],
                providers: [UsMapService],
                exports: [UsMapComponent]
            }]
    }], function () { return []; }, null); })();
    return UsMapModule;
}());
export { UsMapModule };
