// ng-zorro-antd/esm2015/result/result-cells.js
/**
 * @fileoverview added by tsickle
 * Generated from: result-cells.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class NzResultTitleDirective {
}
NzResultTitleDirective.ɵfac = function NzResultTitleDirective_Factory(t) { return new (t || NzResultTitleDirective)(); };
NzResultTitleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzResultTitleDirective, selectors: [["div", "nz-result-title", ""]], hostAttrs: [1, "ant-result-title"], exportAs: ["nzResultTitle"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzResultTitleDirective, [{
        type: Directive,
        args: [{
                selector: 'div[nz-result-title]',
                exportAs: 'nzResultTitle',
                host: {
                    class: 'ant-result-title'
                }
            }]
    }], null, null); })();
export class NzResultSubtitleDirective {
}
NzResultSubtitleDirective.ɵfac = function NzResultSubtitleDirective_Factory(t) { return new (t || NzResultSubtitleDirective)(); };
NzResultSubtitleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzResultSubtitleDirective, selectors: [["div", "nz-result-subtitle", ""]], hostAttrs: [1, "ant-result-subtitle"], exportAs: ["nzResultSubtitle"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzResultSubtitleDirective, [{
        type: Directive,
        args: [{
                selector: 'div[nz-result-subtitle]',
                exportAs: 'nzResultSubtitle',
                host: {
                    class: 'ant-result-subtitle'
                }
            }]
    }], null, null); })();
export class NzResultIconDirective {
}
NzResultIconDirective.ɵfac = function NzResultIconDirective_Factory(t) { return new (t || NzResultIconDirective)(); };
NzResultIconDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzResultIconDirective, selectors: [["i", "nz-result-icon", ""], ["div", "nz-result-icon", ""]], exportAs: ["nzResultIcon"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzResultIconDirective, [{
        type: Directive,
        args: [{
                selector: 'i[nz-result-icon], div[nz-result-icon]',
                exportAs: 'nzResultIcon'
            }]
    }], null, null); })();
export class NzResultContentDirective {
}
NzResultContentDirective.ɵfac = function NzResultContentDirective_Factory(t) { return new (t || NzResultContentDirective)(); };
NzResultContentDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzResultContentDirective, selectors: [["div", "nz-result-content", ""]], hostAttrs: [1, "ant-result-content"], exportAs: ["nzResultContent"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzResultContentDirective, [{
        type: Directive,
        args: [{
                selector: 'div[nz-result-content]',
                exportAs: 'nzResultContent',
                host: {
                    class: 'ant-result-content'
                }
            }]
    }], null, null); })();
export class NzResultExtraDirective {
}
NzResultExtraDirective.ɵfac = function NzResultExtraDirective_Factory(t) { return new (t || NzResultExtraDirective)(); };
NzResultExtraDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzResultExtraDirective, selectors: [["div", "nz-result-extra", ""]], hostAttrs: [1, "ant-result-extra"], exportAs: ["nzResultExtra"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzResultExtraDirective, [{
        type: Directive,
        args: [{
                selector: 'div[nz-result-extra]',
                exportAs: 'nzResultExtra',
                host: {
                    class: 'ant-result-extra'
                }
            }]
    }], null, null); })();
