// ng-zorro-antd/esm5/upload/interface.js
/**
 * @fileoverview added by tsickle
 * Generated from: interface.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
/**
 * @record
 */
export function NzUploadFile() { }
if (false) {
    /** @type {?} */
    NzUploadFile.prototype.uid;
    /** @type {?|undefined} */
    NzUploadFile.prototype.size;
    /** @type {?} */
    NzUploadFile.prototype.name;
    /** @type {?|undefined} */
    NzUploadFile.prototype.filename;
    /** @type {?|undefined} */
    NzUploadFile.prototype.lastModified;
    /** @type {?|undefined} */
    NzUploadFile.prototype.lastModifiedDate;
    /** @type {?|undefined} */
    NzUploadFile.prototype.url;
    /** @type {?|undefined} */
    NzUploadFile.prototype.status;
    /** @type {?|undefined} */
    NzUploadFile.prototype.originFileObj;
    /** @type {?|undefined} */
    NzUploadFile.prototype.percent;
    /** @type {?|undefined} */
    NzUploadFile.prototype.thumbUrl;
    /** @type {?|undefined} */
    NzUploadFile.prototype.response;
    /** @type {?|undefined} */
    NzUploadFile.prototype.error;
    /** @type {?|undefined} */
    NzUploadFile.prototype.linkProps;
    /** @type {?|undefined} */
    NzUploadFile.prototype.type;
    /* Skipping unhandled member: [key: string]: NzSafeAny;*/
}
/**
 * @record
 */
export function NzUploadChangeParam() { }
if (false) {
    /** @type {?} */
    NzUploadChangeParam.prototype.file;
    /** @type {?} */
    NzUploadChangeParam.prototype.fileList;
    /** @type {?|undefined} */
    NzUploadChangeParam.prototype.event;
    /**
     * Callback type.
     * @type {?|undefined}
     */
    NzUploadChangeParam.prototype.type;
}
/**
 * @record
 */
export function NzShowUploadList() { }
if (false) {
    /** @type {?|undefined} */
    NzShowUploadList.prototype.showRemoveIcon;
    /** @type {?|undefined} */
    NzShowUploadList.prototype.showPreviewIcon;
    /** @type {?|undefined} */
    NzShowUploadList.prototype.showDownloadIcon;
}
/**
 * @record
 */
export function ZipButtonOptions() { }
if (false) {
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.disabled;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.accept;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.action;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.directory;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.openFileDialogOnClick;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.data;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.headers;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.name;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.multiple;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.withCredentials;
    /** @type {?|undefined} */
    ZipButtonOptions.prototype.filters;
    /**
     * @param {?} file
     * @param {?} fileList
     * @return {?}
     */
    ZipButtonOptions.prototype.beforeUpload = function (file, fileList) { };
    /**
     * @param {?} item
     * @return {?}
     */
    ZipButtonOptions.prototype.customRequest = function (item) { };
    /**
     * @param {?} file
     * @return {?}
     */
    ZipButtonOptions.prototype.transformFile = function (file) { };
    /**
     * @param {?} file
     * @return {?}
     */
    ZipButtonOptions.prototype.onStart = function (file) { };
    /**
     * @param {?} e
     * @param {?} file
     * @return {?}
     */
    ZipButtonOptions.prototype.onProgress = function (e, file) { };
    /**
     * @param {?} ret
     * @param {?} file
     * @param {?} xhr
     * @return {?}
     */
    ZipButtonOptions.prototype.onSuccess = function (ret, file, xhr) { };
    /**
     * @param {?} err
     * @param {?} file
     * @return {?}
     */
    ZipButtonOptions.prototype.onError = function (err, file) { };
}
/**
 * @record
 */
export function UploadFilter() { }
if (false) {
    /** @type {?} */
    UploadFilter.prototype.name;
    /**
     * @param {?} fileList
     * @return {?}
     */
    UploadFilter.prototype.fn = function (fileList) { };
}
/**
 * @record
 */
export function NzUploadXHRArgs() { }
if (false) {
    /** @type {?|undefined} */
    NzUploadXHRArgs.prototype.action;
    /** @type {?|undefined} */
    NzUploadXHRArgs.prototype.name;
    /** @type {?|undefined} */
    NzUploadXHRArgs.prototype.headers;
    /** @type {?} */
    NzUploadXHRArgs.prototype.file;
    /** @type {?} */
    NzUploadXHRArgs.prototype.postFile;
    /** @type {?|undefined} */
    NzUploadXHRArgs.prototype.data;
    /** @type {?|undefined} */
    NzUploadXHRArgs.prototype.withCredentials;
    /**
     * @param {?} e
     * @param {?} file
     * @return {?}
     */
    NzUploadXHRArgs.prototype.onProgress = function (e, file) { };
    /**
     * @param {?} ret
     * @param {?} file
     * @param {?} xhr
     * @return {?}
     */
    NzUploadXHRArgs.prototype.onSuccess = function (ret, file, xhr) { };
    /**
     * @param {?} err
     * @param {?} file
     * @return {?}
     */
    NzUploadXHRArgs.prototype.onError = function (err, file) { };
}
var NzShowUploadListInterface = /** @class */ (function () {
    function NzShowUploadListInterface() {
    }
    return NzShowUploadListInterface;
}());
export { NzShowUploadListInterface };
