// ng-zorro-antd/esm5/core/element-patch/element-patch.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: element-patch.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive, ElementRef } from '@angular/core';
/**
 * A patch directive to select the element of a component.
 */
import * as ɵngcc0 from '@angular/core';
var NzElementPatchDirective = /** @class */ (function () {
    function NzElementPatchDirective(elementRef) {
        this.elementRef = elementRef;
    }
    /** @nocollapse */
    NzElementPatchDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
NzElementPatchDirective.ɵfac = function NzElementPatchDirective_Factory(t) { return new (t || NzElementPatchDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NzElementPatchDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzElementPatchDirective, selectors: [["", "nz-element", ""]], exportAs: ["nzElement"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzElementPatchDirective, [{
        type: Directive,
        args: [{
                selector: '[nz-element]',
                exportAs: 'nzElement'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
    return NzElementPatchDirective;
}());
export { NzElementPatchDirective };
if (false) {
    /** @type {?} */
    NzElementPatchDirective.prototype.elementRef;
}
